lang/py/avro/schema.py

```diff
@@ -169,7 +170,94 @@ def canonical_properties(self) -> Mapping[str, object]:
         return collections.OrderedDict((key, props[key]) for key in CANONICAL_FIELD_ORDER if key in props)
 
 
-class Schema(abc.ABC, CanonicalPropertiesMixin):
+class FingerprintMixin:
+    """
+    A Mixin to generate schema fingerprints for supported algorithms
+    """
+
+    _FP_TABLE = None
+    _EMPTY64 = 0xC15D213AA4D7A795
+
+    # All algorithms guaranteed by hashlib are supported
+    # Supported Algorithms are:
+    #     - 'blake2b',
+    #     - 'blake2s',
+    #     - 'md5',
+    #     - 'sha1',
+    #     - 'sha224',
+    #     - 'sha256',
+    #     - 'sha384',
+    #     - 'sha3_224',
+    #     - 'sha3_256',
+    #     - 'sha3_384',
+    #     - 'sha3_512',
+    #     - 'sha512',
+    #     - 'shake_128',
+    #     - 'shake_256'
+    algorithms = set(hashlib.algorithms_guaranteed)
+    # Additionally, we provide a custom implementation of 64-bit Rabin fingerprint
+    algorithms.update({"CRC-64-AVRO"})
+    SUPPORTED_ALGORITHMS: FrozenSet[str] = frozenset(algorithms)
+
+    def fingerprint(self, algorithm="CRC-64-AVRO"):
+        """
+        Generate fingerprint for supplied algorithm.
+
+        'CRC-64-AVRO' will be used as the algorithm by default, but any
+        algorithm supported by hashlib (as can be referenced with
+        `hashlib.algorithms_guaranteed`) can be specified.
+
+        `algorithm` param is used as an algorithm name, and NoSuchAlgorithmException
+        will be thrown if the algorithm is not among supported.
+        """
+        schema = self.canonical_form.encode("utf-8")
+
+        if algorithm not in self.SUPPORTED_ALGORITHMS:
+            raise avro.errors.UnknownFingerprintAlgorithmException(f"Unknown Fingerprint Algorithm: {algorithm}")
+
+        if algorithm == "CRC-64-AVRO":
+            return self._crc_64_fingerprint(schema)
+
+        # Generate digests with hashlib for all other algorithms
+        # Lowercase algorithm to support algorithm strings sent by other languages like Java
+        h = hashlib.new(algorithm.lower(), schema)
+        return h.digest()
+
+    def _populate_fp_table(self):
+        self._FP_TABLE = []
+
+        for i in range(256):
+            fp = i
+            for _ in range(8):
+                mask = -(fp & 1)
+                fp = (fp >> 1) ^ (self._EMPTY64 & mask)
+
+            self._FP_TABLE.append(fp)
```

Comment:
It seems that `_FP_TABLE` ends up being completely static after all is said and done. Should it be calculated just once at the module level, at import time?

Reply:
Yeah, I think this whole value can be elevated to the module level and statically written…

```
from functools import reduce
…
_EMPTY64_FINGERPRINT = 0xC15D213AA4D7A795
_FINGERPRINT_TABLE = tuple(reduce(lambda fp, _: (fp >> 1) ^ (_EMPTY64_FINGERPRINT & -(fp & 1)), range(8), i) for i in range(256))
```

What's nice about doing this as a one-liner is that it makes it very clear that this is a completely static value, without having to write it out literally…

…that said, we might still want to put the literal value in a test, and assert that the value calculated here is the same as the one in the test.